### PR TITLE
I 221 add clarification announcement 

### LINKED ISSUE #221: Add a way for a Judge to send a clarification to a team (as well as to All Teams clar)
Thank you for contributing a suggestion for a new feature in PC^2!  
In order to help us understand the proposed feature, please fill in the following fields:

**Is your feature request related to a problem?** Please provide a clear and concise description of what the problem is (for example: "I'm always frustrated when ... " )  
...

**Feature Description**: please provide a clear and concise description of the proposed feature:  
...

**Have you considered other ways to accomplish the same thing?** Please provide a clear and concise description of any alternative solutions or features you've considered.  
...

**Do you have any specific suggestions for how your feature would be ***implemented*** in PC^2?** If so, please describe.  
...

**Additional context**: please provide any additional information that will help us in considering your feature request.  
...

### PR DESCRIPTION
**Pull request by kkarakas**
### Description of what the PR does
Judge can now send announcement clarification to all teams,groups,certain teams with a single button. Their announcement will be displayed as announcement in "All clarifications" tab with correct destinations. In future, how it appears to WTI should be considered.

### Issue which the PR addresses
fixes #221 

### Environment in which the PR was developed (OS,IDE, Java version, etc.)
Windows 11, Eclipse 2021-12 R, JDK 8u381 (1.8)

### Precise steps for _testing_ the PR (i.e., how to demonstrate that it works correctly)
From judge, go to generate clarification, toggle generate announcement. Observe that the text on button changes, title changes (hidden under checkbox right now), various buttons come. When announcement is sent all users receive a clarification without question. 
Judge can also sent clarification without question to groups and some users from the same tab.
Observe that Judges and teams display all clarifications tab correctly. Judges now have the status of ANNOUNCED and shows all the teams,groups it sent.

### CODE REVIEW

File: src/edu/csus/ecs/pc2/core/IInternalController.java [hunk 1]
```diff
@@ -176,7 +178,13 @@ void submitJudgeRun(Problem problem, Language language, SerializedFile mainFile,
      * @param packet
      */
     void sendToTeams(Packet packet);
-
+    
+    /**
+     * Send to groups and or individual teams specified by ultimateDestination.
+     * @param packet
+     * @param ultimateDestination
```
Comment (johnbrvc) [inline src/edu/csus/ecs/pc2/core/IInternalController.java:185]: A comment indicating what this is would be helpful.  That is, what is "ultimateDestination" specifying?

File: src/edu/csus/ecs/pc2/core/PacketHandler.java [hunk 2]
```diff
@@ -1622,15 +1624,26 @@ private void confirmSubmission(Packet packet, ClientId fromId) {
...
+        // Send to judge rather than team
         Packet confirmPacket = PacketFactory.createClarSubmissionConfirm(contest.getClientId(), fromId, clarification);
         controller.sendToClient(confirmPacket);
 
         // Send to clients and other servers
         if (isServer()) {
             controller.sendToJudgesAndOthers(confirmPacket, true);
         }
+        //check if clarification is actually an announcement clarification.
+        if (clarification.getQuestion().equals("") && clarification.isAnswered()) {
```
Comment (johnbrvc) [inline src/edu/csus/ecs/pc2/core/PacketHandler.java:1638]: Maybe make a method in Clarification that tests if a clarification is an announcement by performing this test in the object?

File: src/edu/csus/ecs/pc2/core/PacketHandler.java [hunk 3]
```diff
@@ -1622,15 +1624,26 @@ private void confirmSubmission(Packet packet, ClientId fromId) {
...
+            if (ultimateDestination == null || ultimateDestination.length == 0) {
+                controller.sendToTeams(confirmPacket);//TODO modift to accomodate ultimate destination
```
Comment (johnbrvc) [inline src/edu/csus/ecs/pc2/core/PacketHandler.java:1640]: I don't understand what the "TODO" is for.

File: src/edu/csus/ecs/pc2/core/InternalController.java [hunk 4]
```diff
@@ -2659,7 +2661,80 @@ public void sendToTeams(Packet packet) {
...
+//            abort = false;
+//        }
+//        if (!abort) {
+//            sendPacketToClients(packet, ClientType.Type.TEAM);
+//        }
+        
+        // sendToClient(ConnectionHandlerID connectionHandlerID, Packet packet
+        ClientId[] clientIds = contest.getLocalLoggedInClients(ClientType.Type.TEAM);
+        
+        List<ClientId> badClients = new ArrayList<ClientId>();
```
Comment (johnbrvc) [inline src/edu/csus/ecs/pc2/core/InternalController.java:2694]: Is badClients used anywhere?

File: src/edu/csus/ecs/pc2/core/InternalController.java [hunk 5]
```diff
@@ -2659,7 +2661,80 @@ public void sendToTeams(Packet packet) {
...
+                    sendToClient(connectionHandlerID, packet);
+                }catch (Exception e) {
+                    getLog().log(Level.WARNING, "Exception attempting to send packet using individual team to client " + clientId + "at connectionHandlerId " + connectionHandlerID
+                            + ": " + packet + ": "+ e.getMessage(), e);
+                
+                }
+                continue;
+            }
+            for (IElementObject destination : ultimateDestination) {
+                if (destination instanceof Account) {
```
Comment (johnbrvc) [inline src/edu/csus/ecs/pc2/core/InternalController.java:2718]: From what I can tell in SubmitClarification.Submit(), ultimateDestination can only be a list of group element Ids.  Is that right?   If that is the case, then perhaps a refactor is in order to change the name of ultimateDestination to something like destinationGroups ?

File: src/edu/csus/ecs/pc2/core/model/Clarification.java [hunk 6]
```diff
@@ -136,7 +136,34 @@ public boolean isSendToAll() {
...
+        if (getSubmitter().equals(account.getClientId())){
+            return true;
+        }
+        if (getAnswer() == null) {
+            //THere is no answer to this clar yet hence there are no destinations for it.
+            return false;
+        }
+        ElementId[] destinations = getFirstAnswer().getAllDestinationsGroup();
+        if (destinations != null) {
+            for (ElementId destionation: destinations){
```
Comment (johnbrvc) [inline src/edu/csus/ecs/pc2/core/model/Clarification.java:154]: Spelling error?  destionation -> destination ?

File: src/edu/csus/ecs/pc2/ui/AnswerClarificationPane.java [hunk 7]
```diff
@@ -176,7 +177,7 @@ private JPanel getButtonPane() {
 
     private ClarificationAnswer getClarificationFromFields() {
         ClarificationAnswer clarificationAnswer = new ClarificationAnswer(getAnswerTextArea().getText(), getContest().getClientId(), 
-                getSendToAllCheckBox().isSelected(), getContest().getContestTime());
+                getSendToAllCheckBox().isSelected(), new ElementId[0],getContest().getContestTime());
```
Comment (johnbrvc) [inline src/edu/csus/ecs/pc2/ui/AnswerClarificationPane.java:180]: What is ElementId[0] ?  Comment needed or something.

File: src/edu/csus/ecs/pc2/core/InternalController.java (hunk 4 above)
Comment (kkarakas) [reply src/edu/csus/ecs/pc2/core/InternalController.java:2694]: My understanding is that ClientId's where packet sending has failed must be collected. So I have to use that arraylist somehow.

File: src/edu/csus/ecs/pc2/core/PacketHandler.java (hunk 2 above)
Comment (kkarakas) [reply src/edu/csus/ecs/pc2/core/PacketHandler.java:1638]: Will change.

File: src/edu/csus/ecs/pc2/core/packet/PacketFactory.java [hunk 8]
```diff
@@ -357,6 +357,11 @@ public final class PacketFactory {
 
     public static final String REQUEST_LOGIN_AS_PROXY = "REQUEST_LOGIN_AS_PROXY";
     
+    /**
+     * Ultimate destination on where answers to clarification is sent to
```
Comment (johnbrvc) [inline src/edu/csus/ecs/pc2/core/packet/PacketFactory.java:361]: If these are groups, it should say that.

File: src/edu/csus/ecs/pc2/core/model/Clarification.java (hunk 6 above)
Comment (kkarakas) [reply src/edu/csus/ecs/pc2/core/model/Clarification.java:154]: destination

File: src/edu/csus/ecs/pc2/ui/JCheckBoxJList.java [hunk 9]
```diff
@@ -1,4 +1,4 @@
-// Copyright (C) 1989-2019 PC2 Development Team: John Clevenger, Douglas Lane, Samir Ashoo, and Troy Boudreau.
+// Copyright (C) 1989-2019 PC2 Development Team: John Clevenger, Douglas Lane, Samir Ashoo, and .
```
Comment (johnbrvc) [inline src/edu/csus/ecs/pc2/ui/JCheckBoxJList.java:1]: Looks like you delete Troy by accident - please fix.

Comment (kkarakas) [reply src/edu/csus/ecs/pc2/ui/JCheckBoxJList.java:1]: I don't know why it is there. Sorry Troy

File: src/edu/csus/ecs/pc2/ui/AnswerClarificationPane.java (hunk 7 above)
Comment (kkarakas) [reply src/edu/csus/ecs/pc2/ui/AnswerClarificationPane.java:180]: It should be null or a new constructor is needed.

File: src/edu/csus/ecs/pc2/core/IInternalController.java (hunk 1 above)
Comment (kkarakas) [reply src/edu/csus/ecs/pc2/core/IInternalController.java]: removed ultimateDestination, it now is done by two different arrays one for groups one for teams

File: src/edu/csus/ecs/pc2/core/PacketHandler.java (hunk 2 above)
Comment (kkarakas) [reply src/edu/csus/ecs/pc2/core/PacketHandler.java]: Fixed!

File: src/edu/csus/ecs/pc2/core/PacketHandler.java (hunk 3 above)
Comment (kkarakas) [reply src/edu/csus/ecs/pc2/core/PacketHandler.java]: removed TODO

File: src/edu/csus/ecs/pc2/core/InternalController.java (hunk 5 above)
Comment (kkarakas) [reply src/edu/csus/ecs/pc2/core/InternalController.java]: removed ultimateDestination, it now is done by two different arrays one for groups one for teams

File: src/edu/csus/ecs/pc2/core/packet/PacketFactory.java (hunk 8 above)
Comment (kkarakas) [reply src/edu/csus/ecs/pc2/core/packet/PacketFactory.java]: removed the comments.

File: src/edu/csus/ecs/pc2/ui/AnswerClarificationPane.java (hunk 7 above)
Comment (kkarakas) [reply src/edu/csus/ecs/pc2/ui/AnswerClarificationPane.java]: Constructor added so new array is not needed to initialize.

File: src/edu/csus/ecs/pc2/core/model/Clarification.java (hunk 6 above)
Comment (kkarakas) [reply src/edu/csus/ecs/pc2/core/model/Clarification.java]: fixed!

File: src/edu/csus/ecs/pc2/ui/JCheckBoxJList.java (hunk 9 above)
Comment (kkarakas) [reply src/edu/csus/ecs/pc2/ui/JCheckBoxJList.java:1]: fixed!

File: src/edu/csus/ecs/pc2/core/InternalController.java (hunk 4 above)
Comment (kkarakas) [reply src/edu/csus/ecs/pc2/core/InternalController.java:2693]: Rest is added

**Comment by kkarakas** [comment]
Merge conflicts does not appear to have caused any problems.

**Review by johnbrvc — APPROVED** [review]
Tested at GNY and PacNW Regionals.  Seems to work. There will be a couple of Issue reports following after this PR is merged to address some GUI issues.

File: src/edu/csus/ecs/pc2/core/archive/PacketFormatter.java
Comment (SamanwaySadhu) [inline src/edu/csus/ecs/pc2/core/archive/PacketFormatter.java:1]: Change the copyright year to 2024

File: src/edu/csus/ecs/pc2/core/imports/ContestXML.java
Comment (SamanwaySadhu) [inline src/edu/csus/ecs/pc2/core/imports/ContestXML.java:1]: Change the copyright year to 2024

File: src/edu/csus/ecs/pc2/core/report/ContestAnalysisReport.java
Comment (SamanwaySadhu) [inline src/edu/csus/ecs/pc2/core/report/ContestAnalysisReport.java:1]: Change the copyright year to 2024

File: src/edu/csus/ecs/pc2/services/core/ClarificationJSON.java
Comment (SamanwaySadhu) [inline src/edu/csus/ecs/pc2/services/core/ClarificationJSON.java:1]: Change the copyright year to 2024

File: src/edu/csus/ecs/pc2/services/core/OriginalEventFeedJSON.java
Comment (SamanwaySadhu) [inline src/edu/csus/ecs/pc2/services/core/OriginalEventFeedJSON.java:1]: Change the copyright year to 2024

File: src/edu/csus/ecs/pc2/services/web/ClarificationService.java
Comment (SamanwaySadhu) [inline src/edu/csus/ecs/pc2/services/web/ClarificationService.java:1]: Change the copyright year to 2024

File: src/edu/csus/ecs/pc2/ui/AnswerClarificationPane.java
Comment (SamanwaySadhu) [inline src/edu/csus/ecs/pc2/ui/AnswerClarificationPane.java:1]: Change the copyright year to 2024

File: src/edu/csus/ecs/pc2/ui/ClarificationsPane.java
Comment (SamanwaySadhu) [inline src/edu/csus/ecs/pc2/ui/ClarificationsPane.java:1]: Change the copyright year to 2024

File: src/edu/csus/ecs/pc2/ui/JPanePlugin.java
Comment (SamanwaySadhu) [inline src/edu/csus/ecs/pc2/ui/JPanePlugin.java:1]: Change the copyright year to 2024

File: test/edu/csus/ecs/pc2/services/core/EventFeedJSONTest.java
Comment (SamanwaySadhu) [inline test/edu/csus/ecs/pc2/services/core/EventFeedJSONTest.java:1]: Change copyright year to 2024

**Review by SamanwaySadhu — CHANGES_REQUESTED** [review]
Reviewed only code and not execution. Did not find anything major. Just change the first copyright lines to reflect 2024 to a few files. Approved otherwise.

**Review by SamanwaySadhu — APPROVED** [review]
LGTM!

**Pull request merged**